api.go

```diff
@@ -830,3 +834,111 @@ func (c Client) makeTargetURL(bucketName, objectName, bucketLocation string, que
+	if c.endpointURL.Scheme == "https" && strings.Contains(bucketName, ".") {
+		return false
+	}
+	return c.isVirtualHostStyle
+}
+
+// setVirtualHostStyle - sets the isVirtualHostStyle bool
+// to true if server can accept virtual hosted style urls.
+// Send a HEAD bucket request for a non existent bucket to
+// server with request url in virtual hosted style format,
+// and see if server responds with NotFound status.
+func (c *Client) setVirtualHostStyle() {
+	// set default to false
+	c.isVirtualHostStyle = false
+
+	host := c.endpointURL.Host
+	scheme := c.endpointURL.Scheme
+	method := "HEAD"
+	// Strip port 80 and 443 so we won't send these ports in Host header.
+	if h, p, err := net.SplitHostPort(host); err == nil {
+		if scheme == "http" && p == "80" || scheme == "https" && p == "443" {
+			host = h
+		}
+	}
+	// bucket doesn't exist.
+	bucketName := "auto-probe-bucketname"
+	// Construct HEAD bucket request with virtual style
+	urlStr := scheme + "://" + bucketName + "." + host + "/"
+
+	u, err := url.Parse(urlStr)
+	if err != nil {
+		return
+	}
+	// Initialize a new HTTP request for the method.
+	req, err := http.NewRequest(method, u.String(), nil)
+	if err != nil {
+		return
+	}
+
+	// Get credentials from the configured credentials provider.
+	value, err := c.credsProvider.Get()
+	if err != nil {
+		return
+	}
+
+	var (
+		signerType      = value.SignerType
+		accessKeyID     = value.AccessKeyID
+		secretAccessKey = value.SecretAccessKey
+		sessionToken    = value.SessionToken
+	)
+
+	// Custom signer set then override the behavior.
+	if c.overrideSignerType != credentials.SignatureDefault {
+		signerType = c.overrideSignerType
+	}
+
+	// If signerType returned by credentials helper is anonymous,
+	// then do not sign regardless of signerType override.
+	if value.SignerType == credentials.SignatureAnonymous {
+		signerType = credentials.SignatureAnonymous
+	}
+	// Set 'User-Agent' header for the request.
+	c.setUserAgent(req)
+
+	metadata := requestMetadata{
+		bucketName:       bucketName,
+		contentSHA256Hex: emptySHA256Hex,
+	}
+	// Set all headers.
+	for k, v := range metadata.customHeader {
+		req.Header.Set(k, v[0])
+	}
+
+	req.Body = nil
+	// Set incoming content-length.
+	req.ContentLength = metadata.contentLength
+	switch {
+	case signerType.IsV2():
+		// Add signature version '2' authorization header.
+		req = s3signer.SignV2(*req, accessKeyID, secretAccessKey)
+	default:
+		// Set sha256 sum for signature calculation only with signature version '4'.
+		shaHeader := unsignedPayload
+		if metadata.contentSHA256Hex != "" {
+			shaHeader = metadata.contentSHA256Hex
+		}
+		req.Header.Set("X-Amz-Content-Sha256", shaHeader)
+		// Add signature version '4' authorization header.
+		req = s3signer.SignV4(*req, accessKeyID, secretAccessKey, sessionToken, "us-east-1")
+	}
+	// Initiate the request.
+	r, _ := c.do(req)
+	if r != nil && r.StatusCode == http.StatusNotFound {
```

Comment:
what happens if "auto-probe-bucketname" is
a) exists and owned by a different user in AWS S3 (or GCS etc)
b) exists and owned by this user

Shouldn't we accommodate errors due to above possibilities too?

Reply:
yeah. In case of AWS S3, any user kidding and creating a auto-probe-bucketname bucket in AWS S3 will make all software depending on minio-go switch to use path style. (I am not sure why minio-go is always selecting virtual host style with AWS S3 though)

Reply:
I agree that there is a possibility of preexisting bucket with that name - generating a random bucket name could mitigate possibility of a bucket existing with same name on AWS. Do you think that is better approach?

Reply:
> I agree that there is a possibility of preexisting bucket with that name - generating a random bucket name could mitigate possibility of a bucket existing with same name on AWS. Do you think that is better approach?

I think so.. but the bucket name should be highly random, otherwise it will be dangerous as I explained above.

For example, using 'autoprobe-bucketname' + fmt.Sprintf("%d\", rand.Intn(10000)) won't work because someone can create all the possibility. Then all software using minio-go will get Access denied instead of Bucket Not Found, hence path style will be always activated in case of AWS S3.

Reply:
@vadmeste, updated PR to also check if server responds with StatusOK in the event bucket already exists. StatusOK indicates that server could resolve the url path, and hence can support virtual hosted style url.